Add class generating high-order immersed quadrature rules

simonsticko opened the pull request:
Add a class NonMatching::QuadratureGenerator that generates high-order
immersed quadrature rules over a BoundingBox, B, when the domain is
described by a level set function, \psi. Thus creating quadrature rules
for the following 3 regions of the box:
    
{x \in B : \psi(x) < 0},
{x \in B : \psi(x) > 0},
{x \in B : \psi(x) = 0}.

simonsticko commented:
This is somewhat large, sorry about that...

I'm still working cleaning up the tests (not in the pull-request yet) and on the documentation (especially in `internal::QuadraturGeneratorImplementation`), but I wanted to open the WIP anyway.

If one wants to quickly understand what this is, it is probably easiest to look at
`tests/non_matching/quadrature_generator_sphere.cc`.

Depends on #12015 in order to work.

kronbichler commented:
Thanks @simonsticko for this large PR - I will look into it over the next few days so we can include this (extremely useful) functionality in deal.II at some point.

kronbichler commented on source/non_matching/quadrature_generator.cc:
```diff
@@ -0,0 +1,1364 @@
+            const double   weight = 1;
+            const Point<1> point(roots[i]);
+
+            Tensor<1, 1> normal        = level_sets[0].get().gradient(point);
```
Can you call this `gradient` or something like that? It should be close to a normal vector for a good signed distance function, but as the assert shows we cannot guarantee it here. As a side note, can you think of a case where the hardcoded tolerance `1e-11` below would be bad and we would need a scaling? We often try to make comparisons in the library also work when say the domain has extent 1e20 (or 1e-20). But I guess for a signed distance function we would be really eliminate the h scale, so I think we are fine.

simonsticko replied:
I can change the name, but I normalize the tensor just a few lines below. So are you sure that's what you want?

Regarding the assert, I guess you can always come up with a function where it fails:

\psi(x) = k (x- x_{root}),        k --> 0

The main reason for the assert here is to give a better error message in the case that we try to normalize something that gives a NaN. Given that, I'm not sure what a good value in the assert would be. Do you think it would be better to avoid the division completely with something like this?
```c++
normal[0] = boost::math::sign(gradient[0])
```

kronbichler replied:
I agree with your point - let's just keep these things as they are now. If we ever hit a case with a bad error (and signed distance functions should by construction be scale-invariant), we might search for ideas then.

kronbichler reviewed (approved):
Thank you for this well-structured contribution. I really like how you were able to split the very complicated algorithm by Saye for the quadrature into small pieces such that every piece does a simple task, finally bringing it together.

This is an essential contribution to be able to integrate on arbitrary cuts through the elements (CutFEM), which is something we should strive to support in deal.II.

At some point in the future we might think about the possibility to generalize this (or similar concepts) also to the case for other description of the interface, say an explicit polynomial description of the surface. I believe part of this infrastructure is also useful there, even though the implicit function theorem is (obviously) not needed here.

kronbichler commented:
The indentation script complains about the line breaks in some updated comments, plus this one in the tidy test:
```
/jenkins/workspace/dealii-tidy_PR-12125/source/non_matching/quadrature_generator.cc:1228:9: error: use range-based for loop instead [modernize-loop-convert,-warnings-as-errors]
/jenkins/workspace/dealii-tidy_PR-12125/source/non_matching/quadrature_generator.cc:522:9: error: use range-based for loop instead [modernize-loop-convert,-warnings-as-errors]
```

simonsticko commented on include/deal.II/non_matching/quadrature_generator.h:
```diff
@@ -0,0 +1,1051 @@
+       * N = \{x \in B : \psi_i(x) < 0, \forall i \}, \\
+       * P = \{x \in B : \psi_i(x) > 0, \forall i \}, \\
+       * I = B \setminus (\overline{N} \cup \overline{P}).
+       * @f]
```
I rewrote this slightly, but the changes a rather minor. Please have a look.

simonsticko commented:
I fixed clang-tidy, formatting and squashed the changes.

A few things that I wonder: Are we happy with these function names?

```c++
void
create_quadrature(const Function<dim> &   level_set,
                  const BoundingBox<dim> &box);

const Quadrature<dim> &
get_inside_quadrature() const;

const Quadrature<dim> &
get_outside_quadrature() const;
```
Or would `get_positive_quadrature` /`get_negative_quadrature` be more intuitive? Since that would relates to the    sign of the level set function. On the other hand, positive/negative sounds like it refers to the quadrature weights.
  
Also, since the class is called `QuadratureGenerator`, would it be better to rename `create_quadrature` to `generate()`?

kronbichler commented:
Thank you for pointing out topics of discussion. Let me answer some now:

> Or would get_positive_quadrature /get_negative_quadrature be more intuitive? 

I like your current choice, e.g., `get_inside_quadrature()`, most right now. There is a good documentation for that and it is used in some of the application fields I am active in. I agree with your assessment of `positive/negative_quadrature` that it might rather confuse some people. Also, I think that positive and negative is also putting information that is not essential (the level set sign) into the name, which might get inconsistent if we find ways to define quadrature formulas with different concepts than level set functions (even though we might have different names).

> Also, since the class is called `QuadratureGenerator`, would it be better to rename `create_quadrature` to `generate()`?

I would prefer `generate` in this context, yes.

Since we're talking names, I wonder if you have thought about `UpThroughDimensionCreator`. I did give it some thinking yesterday but couldn't find a better name. (It is only an internal class so not super important.)

simonsticko commented:
> I would prefer `generate` in this context, yes.

Okey, I will rename it.

> Since we're talking names, I wonder if you have thought about `UpThroughDimensionCreator`. I did give it some thinking yesterday but couldn't find a better name. (It is only an internal class so not super important.)

Yes, I really don't like this class name either, but I can't find a name that I like. I thought a bit about `ZeroContourFinder` or something similar, but that only reflects a part of what it does.

kronbichler commented:
> Yes, I really don't like this class name either, but I can't find a name that I like. I thought a bit about ZeroContourFinder or something similar, but that only reflects a part of what it does

I agree, let's keep the current name for now, unless someone else has a good name.

simonsticko commented:
I renamed `create_quadrature(..)` to `generate(..)` in all the classes. Maybe `UpThroughDimensionCreator::generate(..)` is a bit inconsistent with "Creator" in the class name, but I don't really think it's a problem. Otherwise, I think I've addressed everything we have discussed so far.

kronbichler commented:
Thank you for the modifications! I like this. We can tackle further issues in follow-up PRs.
